src/ir/analysis/derive_copy.rs

```diff
@@ -0,0 +1,416 @@
+        }
+    }
+
+    fn initial_worklist(&self) -> Vec<ItemId> {
+        self.ctx.whitelisted_items().iter().cloned().collect()
+    }
+
+    fn constrain(&mut self, id: ItemId) -> ConstrainResult {
+        trace!("constrain: {:?}", id);
+
+        if self.cannot_derive_copy.contains(&id) || self.cannot_derive_copy_in_array.contains(&id) {
+            trace!("    already know it cannot derive Copy");
+            return ConstrainResult::Same;
+        }
+
+        let item = self.ctx.resolve_item(id);
+        let ty = match item.as_type() {
+            Some(ty) => ty,
+            None => {
+                trace!("    not a type; ignoring");
+                return ConstrainResult::Same;
+            }
+        };
+
+        if item.is_opaque(self.ctx, &()) {
+            let layout_can_derive = ty.layout(self.ctx).map_or(true, |l| {
+                l.opaque().can_trivially_derive_copy()
+            });
+            return if layout_can_derive {
+                trace!("    we can trivially derive Copy for the layout");
+                ConstrainResult::Same
+            } else {
+                trace!("    we cannot derive Copy for the layout");
+                self.insert(id);
+                self.insert_array(id)
+            };
+        }
+
+        match *ty.kind() {
+            // Handle the simple cases. These can derive copy without further
+            // information.
+            TypeKind::Void |
+            TypeKind::NullPtr |
+            TypeKind::Int(..) |
+            TypeKind::Float(..) |
+            TypeKind::Complex(..) |
+            TypeKind::Function(..) |
+            TypeKind::Enum(..) |
+            TypeKind::Reference(..) |
+            TypeKind::BlockPointer |
+            TypeKind::Pointer(..) |
+            TypeKind::UnresolvedTypeRef(..) |
+            TypeKind::ObjCInterface(..) |
+            TypeKind::ObjCId |
+            TypeKind::ObjCSel => {
+                trace!("    simple type that can always derive Copy");
+                ConstrainResult::Same
+            }
+
+            TypeKind::Named => {
+                trace!("    Named cannot derive Copy in array");
+                self.insert_array(id)
+            }
+
+            TypeKind::Array(t, len) => {
+                let cant_derive_copy = self.cannot_derive_copy.contains(&t);
+                if  cant_derive_copy {
+                    trace!("    arrays of T for which we cannot derive Copy \
+                            also cannot derive Copy");
+                    self.insert(id);
+                }
+
+                let cant_derive_copy_in_array = self.cannot_derive_copy_in_array.contains(&t);
+                if  cant_derive_copy_in_array {
+                    trace!("    arrays of T for which we cannot derive Copy \
+                            also cannot derive Copy");
+                    self.insert_array(id);
+                }
+                if cant_derive_copy_in_array || cant_derive_copy {
```

Comment:
Rather than checking for both conditions here and then maybe returning if either cannot derive, let's do the check-and-maybe-return dance after each lookup, so that we aren't continuing to do lookups when we already "know" we can stop considering this item.

```rust
let cannot_derive_copy = ...;
if cannot derive_copy {
    ...
    return self.insert(id);
}

let cannot_derive_copy_in_array = ...;
if cannot_derive_copy_in_array {
    ...
    return self.insert_array(id);
}
```